unstructured_ingest/v2/processes/connectors/kafka.py

```diff
@@ -0,0 +1,324 @@
+import base64
+import json
+import socket
+import typing as t
+from dataclasses import dataclass, field
+from pathlib import Path
+
+from pydantic import Field
+
+from unstructured_ingest.error import (
+    SourceConnectionError,
+    SourceConnectionNetworkError,
+)
+
+from unstructured_ingest.v2.interfaces import (
+    AccessConfig,
+    ConnectionConfig,
+    Downloader,
+    DownloaderConfig,
+    DownloadResponse,
+    FileData,
+    FileDataSourceMetadata,
+    Indexer,
+    IndexerConfig,
+    SourceIdentifiers,
+    download_responses,
+)
+
+from unstructured_ingest.v2.logger import logger
+from unstructured_ingest.v2.processes.connector_registry import SourceRegistryEntry
+
+from unstructured_ingest.logger import logger
+from unstructured_ingest.utils.data_prep import batch_generator
+from unstructured_ingest.utils.dep_check import requires_dependencies
+
+if t.TYPE_CHECKING:
+    from confluent_kafka import Consumer, Producer
+
+
+
+class KafkaAccessConfig(AccessConfig):
+    kafka_api_key: t.Optional[str] = Field(description = "Kafka API key to connect at the server")
+    secret: t.Optional[str] = Field(description = "")
+
+
+class KafkaConnectionConfig(ConnectionConfig):
+    bootstrap_server: str
+    port: str
+    topic: str
+    access_config: KafkaAccessConfig
+    confluent: t.Optional[bool] = True
+    num_messages_to_consume: t.Optional[int] = 1
+    timeout: t.Optional[float] = 1.0
+
+
+
+class KafkaIndexerConfig(IndexerConfig):
+    pass
+
+class KafkaIndexer(Indexer):
+    connection_config = KafkaConnectionConfig
+    index_config = KafkaIndexerConfig
+    
+    def initialize(self):
+        topic = self.connection_config.topic
+        logger.info(f"subscribing to topic: {topic}")
+        self.kafka_consumer.subscribe([topic])
+
+    @property
+    def kafka_consumer(self):
+        if self._consumer is None:
+            self._consumer = self.create_consumer()
+        return self._consumer
+
+    @requires_dependencies(["confluent_kafka"], extras="kafka")
+    def create_consumer(self) -> "Consumer":
```

Comment:
This method should live in the connection config class. Similar to many of the other `create_client` methods on other connection configs.